File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        set<char> st;
        int j=0;
        int maxLen =0;
        //sliding window
        for(int i=0;i<s.size();i++)
        {
            while(st.find(s[i]) != st.end())
            {
                st.erase(s[j]);
                j++;
            }
            st.insert(s[i]);
            maxLen = max(maxLen, (int)st.size());
        }
        return maxLen;
        // return st.size();
    }
};

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Map<Character, Integer> lastIndex = new HashMap<>();
        int maxLen = 0;
        int start = 0; // left boundary of current window

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);

            // if character seen and inside current window, move start
            if (lastIndex.containsKey(c) && lastIndex.get(c) >= start) {
                start = lastIndex.get(c) + 1;
            }

            lastIndex.put(c, i); // update last seen index
            maxLen = Math.max(maxLen, i - start + 1);
        }

        return maxLen;
    }
}